pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/readers/StringDictionary.java

```diff
@@ -15,31 +15,71 @@
  */
 package com.linkedin.pinot.core.segment.index.readers;
 
+import com.linkedin.pinot.common.Utils;
 import com.linkedin.pinot.core.segment.index.ColumnMetadata;
 import com.linkedin.pinot.core.segment.memory.PinotDataBuffer;
+import java.io.UnsupportedEncodingException;
 import java.nio.charset.Charset;
+import java.util.Arrays;
+
 
 public class StringDictionary extends ImmutableDictionaryReader {
   private final int lengthofMaxEntry;
   private static Charset UTF_8 = Charset.forName("UTF-8");
   private final char paddingChar;
 
+  // Thread local memory to store byte arrays of strings.
+  private static final ThreadLocal<byte[]> STRING_BYTES_ARRAY = new ThreadLocal<byte[]>() {
+    @Override
+    protected byte[] initialValue() {
+      return null; // Initialized lazily as we don't know the size at this time.
+    }
+  };
+
   public StringDictionary(PinotDataBuffer dataBuffer, ColumnMetadata metadata) {
     super(dataBuffer, metadata.getCardinality(), metadata.getStringColumnMaxLength());
     lengthofMaxEntry = metadata.getStringColumnMaxLength();
     paddingChar = metadata.getPaddingCharacter();
   }
 
   @Override
+  @SuppressWarnings("unchecked")
   public int indexOf(Object rawValue) {
-    final String lookup = rawValue.toString();
-    final int differenceInLength = lengthofMaxEntry - lookup.length();
-    final StringBuilder bld = new StringBuilder();
-    bld.append(lookup);
-    for (int i = 0; i < differenceInLength; i++) {
-      bld.append(paddingChar);
+    final String lookup = (String) rawValue; // This will always be a string
+    byte[] lookupBytes = lookup.getBytes();
+    byte[] dest = getThreadLocalBytes();
+
+    System.arraycopy(lookupBytes, 0, dest, 0, lookupBytes.length);
+    Arrays.fill(dest, lookupBytes.length, dest.length, (byte) paddingChar);
+    String charsetName = "UTF-8";
+
+    try {
+      return stringIndexOf(new String(dest, charsetName));
+    } catch (UnsupportedEncodingException e) {
+      Utils.rethrowException(e);
+    }
+  }
+
+  /**
+   * Helper method to get the thread local bytes to store a string value.
+   * <ul>
+   *   <li> Allocates memory if thread local memory has not been initialized yet, as max length is not known initially.</li>
+   *   <li> Re-allocates memory, if not sufficient. This will lazily expand the size to max across all dictionaries. </li>
+   * </ul>
+   *
+   * @return
+   */
+  private byte[] getThreadLocalBytes() {
+    byte[] dest = STRING_BYTES_ARRAY.get();
+
+    if (dest == null) {
+      dest = new byte[lengthofMaxEntry];
```

Comment:
1. lengthOfMaxEntry is class member, string bytes array is static. Different dictionaries can have different max
2. both if and else blocks are same
3. need synchronization (it's static member). With synchronization it might work but it's ugly.